draw line

gustavoneves12 opened the pull request.

Thomio-Watanabe committed:
Add headers

gustavoneves12 committed:
initial commit
create a function to draw lines
create a ruby to test the plugin

gustavoneves12 committed:
draw the laser scan information
convert laser to point cloud add draw the lines

gustavoneves12 committed:
draw laser line using the laser scan data

gustavoneves12 committed:
set color configurable
create show range mode
check if ranges is not empty and the number of ranges is equal or greater than two

doudou committed:
Merge pull request #1 from Brazilian-Institute-of-Robotics/draw_laser_line

draw laser line

gustavoneves12 committed:
create a class LineHolder to manage the line creation and configuration
refactor the LaserLine plugin
Set the line StateSet to show a more realistic line

gustavoneves12 committed:
create a ruby script to test laser line plugin

doudou commented on viz/LaserLine.cpp:
```diff
@@ -2,44 +2,123 @@
+
+    double angle = laserScan.start_angle; //start angle
+    double angleStep = laserScan.angular_resolution; //angle step
+    double maxRange = laserScan.maxRange / 1000.0; //convert millimeters to meters
```
Low ranges (0 to 5 I think) are error codes in LaserScan. Have a look at the structure.

gustavoneves12 replied:
I looked the structure and you are right, the ranges from 0 to 5 are error codes.

gustavoneves12 replied:
But, we filled this structure using gazebo's data, so I think this error codes don't have effects.

doudou replied:
Yes, but the plugin takes the data structure as input. Handle them or it's going to go booom when in a few months we forgot about it and start using it for other non-gazebo laser data (or if we modify the rock-gazebo plugin to use those).

I.e. this is a vizualizer for a LaserScan structure. Not a visualizer for a LaserScan structure created by gazebo.

doudou commented on viz/LaserLine.cpp:
```diff
@@ -2,44 +2,123 @@
+
+    vertices->clear();
+
+    if (!ranges.empty() && ranges.size() >= 2) {
```
Invert the condition and return false right away, that would make the control flow a lot easier to understand.

doudou commented on viz/LaserLine.cpp:
```diff
@@ -2,44 +2,123 @@
+            /**
+             * Filter occlusion points and points with values equal the max range
+             */
+            if ((showAllLines) || (abs(pt0[0] - pt1[0]) < 0.1 && (pt0[0] <  limit[0] && pt1[0] < limit[0]))) {
```
This is ... weird. If you want to know if pt0 or pt1 are beyond max range, then test for it. Why are you testing only x ? What are you trying to achieve ?

gustavoneves12 replied:
I did a mistake here. I need to test y too.

doudou commented on viz/LaserLine.cpp:
```diff
@@ -2,44 +2,123 @@
+    if (!ranges.empty() && ranges.size() >= 2) {
+
+
+        for (int i = 0; i < ranges.size()-1; i++) {
```
IMO, the function would be a lot cleaner if
- you would just convert all ranges into points, keeping a bitset of which ranges are valid or not
- you would only THEN build the data that LineHolder requires.

Mixing the two requires that you keep a lot of state (and makes you re-compute data twice or more).

gustavoneves12 replied:
Is there some place which data re-computing is happening? Can you explain this solution with more details? What is the problem with the solution right now?

doudou replied:
> What is the problem with the solution right now?

It's messy

DOing
    vector<> points;
    vector<bool> validity;
    computePoints(points, validity);
    createLines(points, validity)

would be clearer. And would make the determination of a valid point easier as well.

In addition, given that you work in pair of points, pt1 is the pt0 for the next loop. So, you compute each range-to-vector3d mapping twice.

gustavoneves12 commented:
I did some changes. I don't know if I got your guidelines properly. Please, check again for me.

doudou commented on viz/LaserLine.cpp:
```diff
@@ -2,44 +2,158 @@
+    double angleStep = laserScan.angular_resolution; //angle step
+
+    //convert max range to Point and convert the value from meters to milimiters
+    Eigen::Vector3d maxPt = Eigen::Vector3d(meters_to_milimeters(laserScan.maxRange), 0.0, 0.0);
```
Didn't you mean millimeters_to_meters ?

gustavoneves12 replied:
Ops!! It is true! I am going to change it and amend the last commit.

doudou commented on viz/LaserLine.cpp:
```diff
@@ -2,44 +2,158 @@
+        points->push_back(osg::Vec3f(pt[0], pt[1], pt[2]));
+
+        //check if range is valid and if the point is less than the limit
+        validity.push_back(laserScan.isRangeValid(ranges[i]) && (pt[0] < limit[0] && pt[1] < limit[1]));
```
I still don't understand the comparison of pt and limit. As I understand it, it is equivalent to

```
if (pt.x < limit.x && pt.y < limit.y)
```

Am I right ?

gustavoneves12 replied:
You right.

doudou replied:
Then what is this test for ? I don't get it ...

doudou replied:
OK. Got it. Much simpler and totally equivalent, since the two vectors are colinear, is

```
(ranges[i] <= maxRange)
```

gustavoneves12 replied:
In the function isRangeValid it is executed this test. But I need to compute the max point (limit is max  point) too, and test if the point is less than the max point.

gustavoneves12 committed:
remove unnecessary configuration in the LineHolder class
create a function to compute the valid points
create a function to compute the lines

gustavoneves12 commented:
@doudou (PING): can you review the code again?

doudou commented on viz/LaserLine.cpp:
```diff
@@ -2,44 +2,154 @@
+        double dtz =  fabs(points[i+0][2] - points[i+1][2]);
+        double dt = sqrt(dtx * dtx + dty * dty + dtz * dtz);
+
+        //if the distance is bigger than max distance, then this line is dropped
```
Finally ... I understand what you are trying to do ... You are filtering points that look like they have nothing "solid"  in-between.

_All_ linear algebra implementations have ways to compute norms. According to [the OSG API documentation](http://trac.openscenegraph.org/documentation/OpenSceneGraphReferenceDocs/a00974.html),

```
double dt = (points[i+0] - points[i+1]).length();
```

gustavoneves12 committed:
use osg vector operator de compute the distance

gustavoneves12 commented:
@doudou PING

doudou commented on manifest.xml:
```diff
@@ -7,13 +7,7 @@
-  -->
   <depend package="gui/vizkit3d" />
   <depend package="base/types" />
+  <rosdep name="osg" />
```
We don't use rosdep tags anymore, only depend tags

doudou commented on viz/LaserLine.cpp:
```diff
@@ -2,44 +2,150 @@
-    base::samples::LaserScan data;
-};
 
+#define milimeters_to_meters(x) (x) / 1000.0
```
just don't use #define in C++.

```
static inline double millimeters_to_meters(double x) { return x / 1000.0; }
```

doudou commented on viz/LaserLine.cpp:
```diff
@@ -2,44 +2,150 @@
 LaserLine::LaserLine()
-    : p(new Data)
 {
+    color = osg::Vec4(0.0, 1.0, 0.0, 0.0);
```
OK. Much better C++ style is to use initialization lists

```
LaserLine::LaserLine()
   : color(osg::Vec4(0.0, 1.0, 0.0, 0.0))
   , vertices(new osg::Vec3Array())
   , ...
```

The reason is that it handles destruction properly if any of the initializations fails.

doudou commented on viz/LaserLine.hpp:
```diff
@@ -3,31 +3,87 @@
+{vizkit3d::Vizkit3DPlugin<base::samples::LaserScan>::updateData(sample);}
+
+private:
+Q_PROPERTY( QColor color READ getColor WRITE setColor )
```
Indentation's fucked up

doudou commented on viz/LineHolder.cpp:
```diff
@@ -0,0 +1,122 @@
+    stateset->setMode(GL_LIGHTING, osg::StateAttribute::OFF);
+
+    // Add a blend function.
+    osg::ref_ptr<osg::BlendFunc> blend ( new osg::BlendFunc ( osg::BlendFunc::ONE, osg::BlendFunc::ONE) );
```
This looks very weird. It's actually related to the discussion we had about references vs pointers vs const and non-const the other day.

You create a smart pointer whose job is to destroy the memory when the object it points to is unused _but_ you never pass that smart pointer. Meaning either OSG has weird smart pointer semantics, or the method you call (which takes a non-const pointer) assumes it's getting ownership of the object _but_ the smart pointer thinks so as well and both will delete the object eventually.

Looking at OSG code, the proper way to do this seems to be to not use a smart pointer at all

```
osg::BlendFunc* blend = new osg::BlendFunc ( osg::BlendFunc::ONE, osg::BlendFunc::ONE);
stateset->setAttributeAndModes ( blend, osg::StateAttribute::ON );
```

doudou commented on viz/LineHolder.cpp:
```diff
@@ -0,0 +1,122 @@
+
+    // Set the hint.
+    osg::ref_ptr<osg::Hint> hint ( new osg::Hint ( GL_LINE_SMOOTH_HINT, GL_NICEST ) );
+    stateset->setAttributeAndModes ( hint.get(), osg::StateAttribute::ON );
```
Same comment as above

gustavoneves12 committed:
remove rosdep from manifest

gustavoneves12 committed:
change the variables initialization
remove #define

gustavoneves12 committed:
remove smart pointers

gustavoneves12 commented:
I used the eclipse auto-indentation to correct the indentation.

doudou commented on viz/LaserLine.cpp:
```diff
@@ -2,44 +2,150 @@
-    : p(new Data)
+    : color(osg::Vec4(0.0, 1.0, 0.0, 0.0))
+    , vertices(new osg::Vec3Array())
+    , lineHolder(new LineHolder())
```
You allocate LineHolder but do not delete it in the destructor. A good practice is to look at all "naked" (i.e. non-smart) pointers declared as attributes of your class and making sure they are (1) initialized to something or NULL in the constructor and (2) deleted in the destructor.

Another question you must ask yourself is: do I need to allocate on the heap ? In this case, I am not sure about it. Allocating "by value" instead of "on the heap" is always safer, do it whenever you can.

gustavoneves12 replied:
I converted the lineHolder to static instance.

gustavoneves12 committed:
convert LineHolder to static instance
fix milimeters_to_meters function

doudou commented:
The code looks good, but "static" has a very specific meaning in C++ (not what you did here). A static instance would be a single object shared between all the instances of the plugin (which would obviously be very bad). What you did is convert it to a value attribute (by opposition to a pointer attribute).
